File: SectionIB/Task65Test.hs
module SectionIB.Task65Test
  ( sectionIBTask65Tests
  , main
  ) where

import SectionIB.Task65Solution (splitPrimeNonPrime)
import TestRunner (Test, assertEqual, runTestMain, testCase, testGroup)

main :: IO ()
main = runTestMain sectionIBTask65Tests

sectionIBTask65Tests :: Test
sectionIBTask65Tests =
  testGroup
    "Section I-b, task 65"
    [ testCase
        "input [] -> expected primes=[], non-primes=[]"
        (assertEqual ([], []) (splitPrimeNonPrime []))
    , testCase
        "input [2,4,3,6,5,9,11] -> expected primes=[2,3,5,11], non-primes=[4,6,9]"
        (assertEqual ([2, 3, 5, 11], [4, 6, 9]) (splitPrimeNonPrime [2, 4, 3, 6, 5, 9, 11]))
    , testCase
        "input [-7,0,1,2,3,-2] -> expected primes=[2,3], non-primes=[-7,0,1,-2]"
        (assertEqual ([2, 3], [-7, 0, 1, -2]) (splitPrimeNonPrime [-7, 0, 1, 2, 3, -2]))
    , testCase
        "input [13,15,13,13] -> expected primes=[13,13,13], non-primes=[15]"
        (assertEqual ([13, 13, 13], [15]) (splitPrimeNonPrime [13, 15, 13, 13]))
    , testCase
        "input [1,8,12,21] -> expected primes=[], non-primes=[1,8,12,21]"
        (assertEqual ([], [1, 8, 12, 21]) (splitPrimeNonPrime [1, 8, 12, 21]))
    ]

File: SectionIB/task65_test.pl
:- module(section_ib_task65_test, [section_ib_task65_tests/1]).

:- use_module('../test_runner').
:- use_module(task65_solution).

:- initialization(main, main).

main :-
    section_ib_task65_tests(Tests),
    run_test_main(Tests).

section_ib_task65_tests(
    test_group('Section I-b, task 65', [
        test_case('input [] -> expected primes=[], non-primes=[]', []-[], Primes1-NonPrimes1, section_ib_task65:split_prime_nonprime([], Primes1, NonPrimes1)),
        test_case(
            'input [11,4,5,0,17,18,-3,19,20,1] -> expected primes=[11,5,17,19], non-primes=[4,0,18,-3,20,1]',
            [11, 5, 17, 19]-[4, 0, 18, -3, 20, 1],
            Primes2-NonPrimes2,
            section_ib_task65:split_prime_nonprime([11, 4, 5, 0, 17, 18, -3, 19, 20, 1], Primes2, NonPrimes2)
        ),
        test_case('input [2] -> expected primes=[2], non-primes=[]', [2]-[], Primes3-NonPrimes3, section_ib_task65:split_prime_nonprime([2], Primes3, NonPrimes3)),
        test_case(
            'input [0,1,-5] -> expected primes=[], non-primes=[0,1,-5]',
            []-[0, 1, -5],
            Primes4-NonPrimes4,
            section_ib_task65:split_prime_nonprime([0, 1, -5], Primes4, NonPrimes4)
        )
    ])
).
